embassy-stm32/src/dma/ringbuffer/mod.rs

```diff
@@ -0,0 +1,293 @@
+    /// If not all of the elements were read, then there will be some elements in the buffer remaining
+    /// The length remaining is the capacity, ring_buf.len(), less the elements remaining after the read
+    /// OverrunError is returned if the portion to be read was overwritten by the DMA controller.
+    pub fn read(&mut self, dma: &mut impl DmaCtrl, buf: &mut [W]) -> Result<(usize, usize), OverrunError> {
+        let readable = self.margin(dma)?.min(buf.len());
+        for i in 0..readable {
+            buf[i] = self.read_buf(i);
+        }
+        let available = self.margin(dma)?;
+        self.read_index.advance(self.cap(), readable);
+        Ok((readable, available - readable))
+    }
+
+    /// Read an exact number of elements from the ringbuffer.
+    ///
+    /// Returns the remaining number of elements available for immediate reading.
+    /// OverrunError is returned if the portion to be read was overwritten by the DMA controller.
+    ///
+    /// Async/Wake Behavior:
+    /// The underlying DMA peripheral only can wake us when its buffer pointer has reached the halfway point,
+    /// and when it wraps around. This means that when called with a buffer of length 'M', when this
+    /// ring buffer was created with a buffer of size 'N':
+    /// - If M equals N/2 or N/2 divides evenly into M, this function will return every N/2 elements read on the DMA source.
+    /// - Otherwise, this function may need up to N/2 extra elements to arrive before returning.
+    pub async fn read_exact(&mut self, dma: &mut impl DmaCtrl, buffer: &mut [W]) -> Result<usize, OverrunError> {
+        let mut read_data = 0;
+        let buffer_len = buffer.len();
+
+        poll_fn(|cx| {
+            dma.set_waker(cx.waker());
+
+            match self.read(dma, &mut buffer[read_data..buffer_len]) {
+                Ok((len, remaining)) => {
+                    read_data += len;
+                    if read_data == buffer_len {
+                        Poll::Ready(Ok(remaining))
+                    } else {
+                        Poll::Pending
+                    }
+                }
+                Err(e) => Poll::Ready(Err(e)),
+            }
+        })
+        .await
+    }
+
+    fn update_dma_index(&mut self, dma: &mut impl DmaCtrl) {
+        self.write_index.dma_sync(self.cap(), dma)
+    }
+
+    fn read_buf(&self, offset: usize) -> W {
+        unsafe {
+            core::ptr::read_volatile(
+                self.dma_buf
+                    .as_ptr()
+                    .offset(self.read_index.as_index(self.cap(), offset) as isize),
+            )
+        }
+    }
+
+    /// Returns available dma samples
+    fn margin(&mut self, dma: &mut impl DmaCtrl) -> Result<usize, OverrunError> {
+        self.update_dma_index(dma);
+
+        let diff: usize = self
+            .write_index
+            .diff(self.cap(), &mut self.read_index)
+            .try_into()
+            .unwrap();
```

Comment:
The implicit invariant is that the read index is always behind the write index. It should not happen since we only update the read index up to the diff between it and the write index.